### DISCUSSION TITLE
Host code generation

### DISCUSSION THREAD

**Discussion by Bastacyclop** [rise-lang/shine#96] [RFC]
This discussion is about design choices for the host code generation being implemented in #92. Feel free to comment and propose alternatives.

At the Rise level, values can be computed on a target processor via an OpenCL kernel call by using a new `oclRun` primitive as in this example:

```scala
test("two kernel calls") {
  val e = depFun((n: Nat) => fun((n`.`i32) ->: (n`.`i32))(in =>
    oclRun(LocalSize(2), GlobalSize(n/2))(mapGlobal(add(li32(1)))(in)) |> store(x =>
    oclRun(LocalSize(4), GlobalSize(n/2))(mapGlobal(add(li32(2)))(x)))
  ))
  dumpModule(gen.opencl.hosted.fromExpr(e))
}
```

Which currently generates the code below. Each `oclRun` primitive is replaced by a kernel call on the host side, and the required kernels are generated separately.

```c
__kernel __attribute__ ((reqd_work_group_size(2, 1, 1)))
void k0(global int* restrict output, int n1, const global int* restrict e2){
  /* Start of moved local vars */
  /* End of moved local vars */
  /* mapGlobal */
  for (int gl_id_138 = get_global_id(0);(gl_id_138 < n1);gl_id_138 = (gl_id_138 + (n1 / 2))) {
    output[gl_id_138] = (((int)1) + e2[gl_id_138]);
  }
}
```
```c
__kernel __attribute__ ((reqd_work_group_size(4, 1, 1)))
void k1(global int* restrict output, int n1, const global int* restrict e96){
  /* Start of moved local vars */
  /* End of moved local vars */
  /* mapGlobal */
  for (int gl_id_157 = get_global_id(0);(gl_id_157 < n1);gl_id_157 = (gl_id_157 + (n1 / 2))) {
    output[gl_id_157] = (((int)2) + e96[gl_id_157]);
  }
}
```
```c
#include <stdint.h>

void foo(int32_t* output, int n1, int32_t* e2){
  {
    int32_t x117[n1];
    k0(2, 1, 1, (n1 / 2), 1, 1, x117, n1, e2);
    k1(4, 1, 1, (n1 / 2), 1, 1, output, n1, x117);
  }
}
```

Right now, a kernel call is generated as a plain function call and proper OpenCL host code generation still needs to be implemented.

The planned design involves:
- introducing a new array type in DPIA which can be used both for target and host processing
- this new array type will have an associated lightweight runtime representation and will also be used for the host input/output signature; allowing to receive properly allocated memory (e.g. zero-copy compatible on integrated processors) and to avoid potentially expensive conversion operations.
- for prototype simplicity, this array type will not be visible in Rise and when lowering to DPIA we will need a pass to introduce such arrays as well as synchronization points which can either trigger memory copies between host and target, or trigger zero-copy memory mappings
- the host function be given an OpenCL context by argument
- in a first step, OpenCL kernel objects will be created on every kernel call; in the future we want to optimize kernel creation and destruction

Here is what the previous `foo` function could look like following this design (half pseudo code and ignores error-handling):
```c
#include <stdint.h>
#include "Runtime.h"

void launch_k0(Context* ctx, Buffer* x117, int n1, Buffer* e2) {
    Kernel* k = loadKernel("k0", k0_path, ctx);
    startBufferSync(ctx, x117, TARGET_WRITE);
    startBufferSync(ctx, e2, TARGET_READ);

    // TODO: should this be OpenCL-agnostic?
    clSetKernelArg(k, 0, sizeof(cl_mem), x117);
    clSetKernelArg(k, 1, sizeof(int), n1);
    clSetKernelArg(k, 2, sizeof(cl_mem), e2);
    size_t local_size[3] = { 2, 1, 1 };
    size_t global_size[3] = { (n1 / 2), 1, 1 };
    clEnqeueNDRangeKernel(ctx, k, 3, NULL, global_size, local_size, 0, NULL, NULL);

    endBufferSync(ctx, x117, TARGET_WRITE);
    endBufferSync(ctx, e2, TARGET_READ);
    destroyKernel(ctx, k);
}

void launch_k1(Context* ctx, Buffer* output, int n1, Buffer* x117) {
    Kernel* k = loadKernel("k1", k1_path, ctx);
    startBufferSync(ctx, output, TARGET_WRITE);
    startBufferSync(ctx, x117, TARGET_READ);

    // TODO: should this be OpenCL-agnostic?
    clSetKernelArg(k, 0, sizeof(cl_mem), output);
    clSetKernelArg(k, 1, sizeof(int), n1);
    clSetKernelArg(k, 2, sizeof(cl_mem), x117);
    size_t local_size[3] = { 4, 1, 1 };
    size_t global_size[3] = { (n1 / 2), 1, 1 };
    clEnqeueNDRangeKernel(ctx, k, 3, NULL, global_size, local_size, 0, NULL, NULL);

    endBufferSync(ctx, output, TARGET_WRITE);
    endBufferSync(ctx, x117, TARGET_READ);
    destroyKernel(ctx, k);
}

void foo(Context* ctx, Buffer* output, int n1, Buffer* e2){
  {
    Buffer* x117 = createBuffer(ctx, n1 * sizeof(int32_t), TARGET_READ | TARGET_WRITE);
    launch_k0(ctx, x117, n1, e2);
    launch_k1(ctx, output, n1, x117);
    destroyBuffer(ctx, x117);
  }
}
```
```c
#define CL_TARGET_OPENCL_VERSION 120
#define CL_USE_DEPRECATED_OPENCL_1_2_APIS
#ifdef __APPLE__
    #include "OpenCL/opencl.h"
#else
    #include "CL/cl.h"
#endif

// example, different for CUDA
struct Context {
  cl_context ocl;
};

// example, different for zero-copy
struct Buffer {
  void* host_ptr;
  cl_mem target_mem;
};

Buffer* createBuffer(Context* ctx, size_t byte_size, AccessFlags access) {
  Buffer* b = malloc(sizeof(Buffer));
  if (contains(HOST_READ, access) || contains(HOST_WRITE, access)) {
    b->host_mem = malloc(byte_size);
  } else {
    b->host_mem = NULL;
  }
  if (contains(TARGET_READ, access) || contains(TARGET_WRITE, access)) {
    b->target_mem = clCreateBuffer(ctx->ocl, accessToOclFlags(access), byte_size, NULL, &ocl_err);
  } else {
    b->target_mem = 0;
  }
}

void destroyBuffer(Context* ctx, Buffer* b) {
  if (b->host_ptr != NULL) { free(b->host_ptr); }
  if (b->target_mem != 0) { clReleaseMemObject(b->target_mem); }
  free(b);
}

void startBufferSync(Context* ctx, Buffer* b, AccessFlags access) {
  // TODO
}

void endBufferSync(Context* ctx, Buffer* b, AccessFlags access) {
  // TODO
}

// -- zero copy variants --

struct Buffer {
  cl_mem mem;
};

Buffer* createBuffer(Context* ctx, size_t byte_size, AccessFlags access) {
  Buffer* b = malloc(sizeof(Buffer));
  b->mem = clCreateBuffer(ctx->ocl, CL_MEM_ALLOC_HOST_PTR | accessToOclFlags(access), byte_size, NULL, &ocl_err);
}

void destroyBuffer(Context* ctx, Buffer* b) {
  clReleaseMemObject(b->mem);
  free(b);
}

void startBufferSync(Context* ctx, Buffer* b, AccessFlags access) {
  // TODO, might need to clEnqueueMapBuffer and return a host pointer
}

void endBufferSync(Context* ctx, Buffer* b, AccessFlags access) {
  // TODO, might need to clEnqueueUnmapMemObject
}

```

Bag of ideas:
- We could use address spaces to differentiate `OpenCL` memory, `Stack` memory and `Heap` memory on the host. However we prefer to keep a simpler design in a first prototype.

**Comment by fedepiz** [comment]
So, read it through, and looks fine to me. I do have a general question/suggestion/observation though.

Is adding a new type of array at the DPIA level the right conceptual design? We already have two array types ("normal" and "position-dependent"), which differ in behaviour but ultimately share the low-level representation. 
Is there way to be composable about it, at least in principle? Otherwise, we risk an explosion of types.

(Self counterargument: This may be "putting to much meat on the table", and perhaps it is better to start with the simpler, more straightforward design. Then generalise later.
Self counterargument 2: How many different complex types we will ever need eventually? Arrays, Dependent Arrays, Pairs of Arrays, Dependent Pairs are already in...is it worth the extra complexity just on the off-chance somebody, down the line, starts adding many new types?)

PS - I have no concrete suggestions on how to go about it :D This is 100% pure naval-gazing on my part

**Reply by Bastacyclop** [reply]
You could have a tuple like `(array in target address space, array in host address space)` and then projections and syncs. However this is not the desired implementation on integrated processors with zero-copy possibility. The advantage of having a new "opaque" type is that we can delegate some implementation choices to a lightweight runtime.

**Comment by Bastacyclop** [comment]
Here is an example of code that is generated right now for a pure host-side identity:
```scala
  test("no kernel call") {
    val e = depFun((n: Nat) => fun((n`.`i32) ->: (n`.`i32))(in =>
      mapSeq(add(li32(1)))(in)
    ))
    dumpModule(gen.opencl.hosted.fromExpr(e))
  }
```
```c
void foo(Context ctx, Buffer moutput, int n1, Buffer me2){
  {
    int32_t* e2 = ((int32_t*)hostBufferSync(ctx, me2, (4*n1), HOST_READ));
    int32_t* output = ((int32_t*)hostBufferSync(ctx, moutput, (4*n1), HOST_WRITE));
    /* mapSeq */
    for (int i_35 = 0;(i_35 < n1);i_35 = (1 + i_35)) {
      output[i_35] = (((int32_t)1) + e2[i_35]);
    }
  }
}
```

**Comment by Bastacyclop** [comment]
Given an OpenCL module named "foo" with 2 kernels I was thinking of generating:
- `foo/host.c`
- `foo/k0.cl`
- `foo/k1.cl`

Now `host.c` will depend on a runtime header, a runtime implementation and the path to `foo/` so that it can load its kernels at runtime. How can we facilitate compilation and runtime execution of this module? Or maybe more relevant short-term: how would we automate executing a function from this module from our unit tests, where we might want to write the test using plain C code which calls the generated function like we often already do?

**Reply by fedepiz** [reply]
Assuming we know the name of `void foo` from scala (maybe pass it down to dumpModule? Or use a fixed name/defined symbol?), and the path to `foo/` (which we really do :D), then I can't see something much simpler then
1) Provide a scala String with the source of a C driver source, with `main` in it, from which `void foo()` can be called.
2) Save the source in `main.c`
2) Compile `main.c`, `host.c`, and link them together with the runtime to make an executable
3) Invoke the binary

There may be some fiddlyiness around the paths to `foo` for the OpenCL sources, but I think the "obvious way" will do.

Any clear downside of this ultra-naive approach, at least for starters?

**Reply by Bastacyclop** [reply]
That sounds reasonable for unit testing and starting out. We also need to pick a runtime implementation but we can simply have an `enum` to pick between different built-ins.
I think that there might be addtionnal questions outside of the unit test world (e.g. I want to take the generated code and use it in my own project which has its own Makefiles) but we can wait until this is required.
